### ISSUE TITLE
Browse all server population issues

### ISSUE THREAD

**Issue by pamduncan** [StrangeLoopGames/EcoIssues#10131]
There have been complaints that few to no  servers are showing up for "browse all" .  When you first go into "Browse all"   only about 10 servers show up.  You can go into filters and uncheck 'version match' and 'language match' and that will populated approximately 20 servers .  If you then do a empty search it will populate a great many more at that point but still not all servers.  Sometimes a owners server will be in this larger list, other times it will not and the majority of these cases are with servers that have been up, online and showing in 'browse all' for several days at least.

Deleting appdata and allowing it to recreate did not change what is happening.

I have been able to repro this on a on a few different connections to the game

People are also not seeing their server on in the Browse all menu.  Sometimes it will show up if searched for and others cases it does not..   Do not think this is a configuration issue on the server but that the browser is not populating the "browse all" correctly as many of the ones now not showing up are servers that have been showing up for days since the issue was resolved.

 Seems like the server will show up most of the time in the recommended windows if the server was configured with a server category and its a close match to the server owner.  That its not showing up does not mean its not there but just that the owner is not a good match for the server.  I was able to see their server in the recommended list that they could not and they were able to spot servers belonging to others that were not being seen by their owner or me.  So not showing up in the Recommended list is not a real check point unless the owners has been able to see his server on the recommended list up until now.

**Comment by pamduncan** [comment]
Thank you Mirasrael for getting this fixed.  People are now reporting that the browser is populating with having to play with filters and doing empty searches to see some servers.   Also between this and shorting up the number of characters used in values in the Network.eco servers that were not showing up are again showing.

Unfortunately  some players are now reporting that the recommended windows are no longer populating at all and are empty.   I am not able to reproduce this as all mine appear to populate just fine.  I have asked them to delete appdata and see if that fixed it.  I am waiting to hear back

**Comment by Nicolausi113** [comment]
Maybe a hint to this problem:
The client v7.7.2 is a quit stable version and worked fine for me, now it doesn’t show any server, even with all filters unchecked.
If I first select Your-Worlds after starting, the few favorites I marked showing all as online (but wrong version). If I then select New-Game, make a server search and go back to Your-Worlds, all favorites now show as offline – curious?

**Comment by Nicolausi113** [comment]
Some more info: v7.8.5 shows some server – and then after 1 minute
![newgame 2](https://user-images.githubusercontent.com/38510829/49022608-8e065180-f195-11e8-8e3d-8149fb8bd7d7.jpg)

**Comment by mirasrael** [comment]
Let's recheck with 7.8.6 release

**Comment by Nicolausi113** [comment]
I’ve updated to 7.8.6, unfortunately the problem still persist. My favorite server doesn’t show up in Recommended nor in Browse All and e.g. the server “Der Rentner Treff” is not shown in Browse ALL
![server1](https://user-images.githubusercontent.com/38510829/50045977-4c820b80-009c-11e9-90c9-e77e934aca1f.jpg)
![server2a](https://user-images.githubusercontent.com/38510829/50045980-5b68be00-009c-11e9-8ed2-946137a6466d.jpg)
![server3a](https://user-images.githubusercontent.com/38510829/50046003-9f5bc300-009c-11e9-9aaf-b08c97b98a79.jpg)

**Comment by mirasrael** [comment]
Hi @Nicolausi113 
I checked search and it was shown "Der Rentner Treff" for me. Is everything fine now?

**Comment by Nicolausi113** [comment]
@mirasrael Thanks for replying! I gave up playing ECO 6 weeks ago and not familiar with current state. If v8.0 is released, I think I’ll try it once more and will report issues if experienced.

**Comment by jthill** [comment]
I'm getting a closely-related symptom on the linux client, installed from the website's `.zip`: intermittent and variably-persistent failures to populate the server lists at all, with one [Access denied] error per "New Game" entry showing in Player.log.  
```
Failed to send packet: System.Net.Sockets.SocketException (0x80004005): Access denied
  at System.Net.Sockets.Socket.SendTo (System.Byte[] buffer, System.Int32 offset, System.Int32 size, System.Net.Sockets.SocketFlags socketFlags, System.Net.EndPoint remoteEP) [0x00082] in <ac210d81537245bc838518cc8e845861>:0 
  at Lidgren.Network.NetPeer.SendPacket (System.Int32 numBytes, System.Net.IPEndPoint target, System.Int32 numMessages, System.Boolean& connectionReset) [0x00028] in <fe68c86a69c94db5b66850721edd94a4>:0 
 
(Filename: ./Runtime/Export/Debug.bindings.h Line: 43)

```

I can reliably produce a related symptom with the "Your Worlds" screen.

So, the intermittent part is simply that it'll go for a day or days without happening, then it'll start happening and persist for an unpredictable length of time, second to hours, then vanish, but ignorantly farting around led me to a couple observations:

* the 32-bit client virtually always successfully populates the list even when the 64-bit client is being stubborn, and bouncing out of the 32-bit client and back to the 64-bit client seems to often fix it.

* running a local server (I know the linux server's unsupported, but this client symptom might be helpful anyway), when I go to "Your Worlds", the server usually shows up in the LAN section on first entry regardless of whether the New Game server list is getting populated, but if I esc out and reenter, the server almost always _doesn't_ show up in the LAN section and shows grayed in the recent list, and the error does pop in the logfile. Unless I mark it as a favorite, then it shows up on reentry active at the top and in the recent list, but doesn't show in the LAN section.

When I did the install I took a git snapshot immediately after extracting, so I can verify the install, there are no differences except the expected ones for identification / setup in the `.eco`'s.  There's no AV involved, the firewall's manually configured and tested, ports 2999-3003 are forwarded unchanged to my box on TCP/UDP, the symptoms come and go without changing anything.

**Comment by mirasrael** [comment]
@jthill Hi, thanks for info. Seems your network security system on Linux for some reason blocks traffic. Client makes a request for every server from list for ping. So seems like these queries treated as unathorized for some reason.

**Comment by jthill** [comment]
Googling for the error message got me to [this answer](https://answers.unity.com/questions/248494/socket-exception-access-denied.html#) saying to enable broadcasting before trying it, and stracing the Eco client and hunting backward from where it wrote the error message got me to this:

> `[pid  3630] sendto(40, "\210\0\0\0\0", 5, 0, {sa_family=AF_INET, sin_port=htons(3000), sin_addr=inet_addr("192.168.254.255")}, 16) = -1 EACCES (Permission denied)`

which is the (correct) broadcast address for the local net, so I think you're not setting the broadcast option on the socket.

**Comment by mirasrael** [comment]
Seems this problem isn't broadcast related. We are using Lidgren.Network library, I checked its code and
```csharp
internal void SendPacket(
      int numBytes,
      IPEndPoint target,
      int numMessages,
      out bool connectionReset)
{
        // skipped
        ipAddress = NetUtility.GetCachedBroadcastAddress();
        if (target.Address == ipAddress)
          this.m_socket.SetSocketOption(SocketOptionLevel.Socket, SocketOptionName.Broadcast, true);
        // send packet, skipped
        if (target.Address == ipAddress)
          this.m_socket.SetSocketOption(SocketOptionLevel.Socket, SocketOptionName.Broadcast, false);
}
```
So if target address is Broadcast then it enabled Broadcast and then disables it afterwards.

**Comment by jthill** [comment]
> @jthill May be problem related to Linux permissions for Broadcast address which app can't overcome?

I no longer think the broadcast problem and the server-list population-problem are related, see below, but strace says no socket's never being enabled for broadcast, to get that with your posted code that cached broadcast address has to be the wrong one for the link, IPv6 is also enabled on the local net so maybe it's getting confused.

But for the real problem, with the 32-bit client the server-list-population problem happens rarely if ever (I don't recall it ever happening), with the 64-bit client it's as described, a crapshoot and variably persistent.

Perhaps more interestingly, when it does happen, exiting the 64-bit client, starting the 32-bit client and looking at the server list, then exiting that and starting the 64-bit client again pretty much always corrects the symptom, at least for a while.  If this were a permission issue, I'd expect it to always happen, and certainly not to change from one run to the next.

But you guys are heads down on 8.0 and there's nothing but an error message behind the download link for the 8.0 linux client, so I'm thinking I'll just go do other things until your team's got the time to spare for this, I do understand about things being harder than they looked at first and not having the resources to cover them.

**Comment by D3nnis3n** [comment]
Is this still recent?

**Comment by jthill** [comment]
This problem looks fixed, just fetched 0.8.0.5, the server list populated properly in `Eco.x86_64`, tried four-five times now across several minutes, that worked every time I tried.
